### ISSUE TITLE
ASC_MEMORY_BASE default value

### ISSUE THREAD

**Issue by lexaknyazev** [AssemblyScript/assemblyscript#2124]
When compiled without explicit `--memoryBase` option, these two functions return different values.
```ts
export function foo(): i32 {
  return ASC_MEMORY_BASE; // returns 0
}

export function bar(): i32 {
  return <i32>memory.data(1); // returns 1024 or 8 (with --lowMemoryLimit)
}
```

It would be more convenient if the value of `ASC_MEMORY_BASE` matched the actual memory base (derived from `--memoryBase` and/or `--lowMemoryLimit` options) or if another compiler-provided const value with the actual memory base was defined.

**Comment by MaxGraey** [comment]
`ASC_MEMORY_BASE` is actual value of external options. But real depends on optimization flags this value could be changed. [I recommend use `__heap_base`](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKZXhwb3J0IGZ1bmN0aW9uIG1lbWVvcnlCYXNlKCk6IGkzMiB7CiAgcmV0dXJuIEFTQ19NRU1PUllfQkFTRTsgLy8gMAp9CgpleHBvcnQgZnVuY3Rpb24gc29tZVN0YXRpY01lbUFsbG9jKCk6IHVzaXplIHsKICByZXR1cm4gbWVtb3J5LmRhdGEoOCk7IC8vIDEwMjQKfQoKZXhwb3J0IGZ1bmN0aW9uIGhlYXBCYXNlKCk6IHVzaXplIHsKICByZXR1cm4gX19oZWFwX2Jhc2U7ICAgIC8vIDEwMjQgKyA4Cn0KIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKQogICAgZm9yIChsZXQgaSA9IDA7IGkgPD0gMTA7ICsraSkgewogICAgICBvdXRwdXQudmFsdWUgKz0gYGZpYigke2l9KSA9ICR7ZXhwb3J0cy5maWIoaSl9XG5gCiAgICB9CiAgfSkKPC9zY3JpcHQ+).

**Comment by lexaknyazev** [comment]
The heap base serves slightly different purpose and isn't sound enough when the heap is not used.

The request is motivated by the following test that guarantees that static blocks start at `ASC_MEMORY_BASE`:
https://github.com/AssemblyScript/assemblyscript/blob/1510bac5d8c69bb266eb5c52248799e989497060/tests/compiler/memorybase.json#L1-L5

https://github.com/AssemblyScript/assemblyscript/blob/1510bac5d8c69bb266eb5c52248799e989497060/tests/compiler/memorybase.ts#L1-L4

**Comment by MaxGraey** [comment]
Another option is using `__data_end`. Which should reflect actual value of `ASC_MEMORY_BASE`

**Comment by lexaknyazev** [comment]
`__data_end` returns 1028 here:
```ts
export function foo(): i32 {
  return <i32>__data_end; // 1028
}

export function bar(): i32 {
  return <i32>memory.data(1); // 1024
}
```

**Comment by MaxGraey** [comment]
Yeah, I agree we should have some global var for actual value of `ASC_MEMORY_BASE`. We need to think about the best way to do it. For now you could do something like this:

```ts
function memoryBase(): usize {
  if (ASC_OPTIMIZE_LEVEL >= 3) {
    return max(ASC_MEMORY_BASE, 1024);
  } else {
    return max(ASC_MEMORY_BASE, 16);
  }
}
```
